### ISSUE TITLE
Nested Dict Spaces

### ISSUE THREAD

**Issue by prototact** [tkn-tub/ns3-gym#58]
Hi,
Is it possible to nest Dict Spaces? It seems to me that they are automatically flattened.

**Comment by Abdelkader-gnichi** [comment]
hi @prototact did manage to make it with  multi-dimensional ?

**Comment by prototact** [comment]
I had to modify the python code but it worked yes. I can share later if you'd like.

**Comment by pgawlowicz** [comment]
Hi @prototact, could you provide a patch for the ns3gym?

**Comment by Abdelkader-gnichi** [comment]
@prototact yes please can you share your modification cos i need it this period to complete my master project and thank you a lot for your response

**Comment by Abdelkader-gnichi** [comment]
Hi @pgawlowicz i hope that you doing well, 

Did you have any implementations or small complet exemple of multi agent learning in ns3-gym, and if you have an example about ns3-gym with ofswitch13 package cos my project is about configure a bunch of switches that receiving udp traffic to make it reach its final destination based on DRL agent decisions (the decisions is an output port for each switch that have an incoming traffic on its ports) 

Again i want to thank you about your massive work on the ns3-gym package so god bless you man.

**Comment by prototact** [comment]
> Hi @prototact, could you provide a patch for the ns3gym?

I made a #88 request, I hope it is okay with you. I should specify that my patch is about supporting nested dicts and not multidimensional arrays, but a similar solution may work (recursively call `_pack_data`)

**Comment by Abdelkader-gnichi** [comment]
@prototact thank you a lot i really appreciate your response and your work.
Can you tell me please about your project that you use ns3-gym to make it.

**Comment by prototact** [comment]
I use 5g-Lena with ns3-gym to simulate 5g network slicing. However, ns3 is still lacking Edge and Core Network support, and it does not look like it will be done any time soon, so I am uncertain if I will proceed with the project.

**Comment by Abdelkader-gnichi** [comment]
Nice project and ns3 it is huge and need time and effort to work with it, my project is a self aware sdn topology that that configure ofswitch13 switches in order to make the traffic reach it's final destination through this switches based on DRL agent decisions (the decisions are the outgoing output port that will be used by the switche\s )

**Comment by Abdelkader-gnichi** [comment]
@prototact I want to ask about feeding the agent with the observation, so did you use OpenGymBoxContainer or OpenGymDict 
And for the actionspace what did you use cos my action space must be discrete ( i need to get 3 value from the agent and use np.argmax to take the the biggest one) so if you have some ideas i will so thankful if you share them with and thanks in advance.

**Comment by Abdelkader-gnichi** [comment]
> Hi @prototact, could you provide a patch for the ns3gym?

Hi @pgawlowicz did you merge the patch of multidimensional arrays in ns3gym?

**Comment by Abdelkader-gnichi** [comment]
Hi @prototact does your patch of ns3-gym sends multiple actions to the simulation part as multi-dict space ?
